inoculate/src/trace.rs

```diff
@@ -0,0 +1,168 @@
+    level: Level,
+    writer: Writer<'writer>,
+    is_empty: bool,
+    color: bool,
+    did_cargo_format: bool,
+}
+
+impl<'writer> Visitor<'writer> {
+    const MESSAGE: &'static str = "message";
+    const INDENT: usize = 12;
+
+    fn new(level: Level, writer: Writer<'writer>, color: bool) -> Self {
+        Self {
+            level,
+            writer,
+            is_empty: true,
+            did_cargo_format: false,
+            color,
+        }
+    }
+}
+
+impl Visit for Visitor<'_> {
+    fn record_debug(&mut self, field: &Field, value: &dyn fmt::Debug) {
+        if self.is_empty {
+            if self.level >= Level::INFO && field.name() == Self::MESSAGE {
+                let message = format!("{:?}", value);
+                if let Some((tag, message)) = message.as_str().split_once(' ') {
+                    if tag.len() <= Self::INDENT && tag.ends_with("ing") || tag.ends_with('d') {
+                        let tag = tag.to_title_case();
+                        let style = match (self.level, self.color) {
+                            (Level::DEBUG, true) => style().bright_blue().bold(),
+                            (_, true) => style().green().bold(),
+                            (_, false) => style(),
+                        };
+
+                        let _ = write!(
+                            self.writer,
+                            "{:>indent$} ",
+                            tag.style(style),
+                            indent = Self::INDENT
+                        );
+
+                        let _ = self.writer.write_str(message);
+                        self.is_empty = false;
+                        self.did_cargo_format = true;
+                        return;
+                    }
+                }
+            }
+
+            let _ = match (self.level, self.color) {
+                (Level::ERROR, true) => {
+                    write!(self.writer, "{}{} ", "error".red().bold(), ":".bold())
+                }
+                (Level::WARN, true) => {
+                    write!(self.writer, "{}{} ", "warning".yellow().bold(), ":".bold())
+                }
+                (Level::INFO, true) => {
+                    write!(self.writer, "{}{} ", "info".green().bold(), ":".bold())
+                }
+                (Level::DEBUG, true) => {
+                    write!(self.writer, "{}{} ", "debug".blue().bold(), ":".bold())
+                }
+                (Level::TRACE, true) => {
+                    write!(self.writer, "{}{} ", "trace".purple().bold(), ":".bold())
+                }
+                (Level::ERROR, false) => self.writer.write_str("error: "),
+                (Level::WARN, false) => self.writer.write_str("warning: "),
+                (Level::INFO, false) => self.writer.write_str("info: "),
+                (Level::DEBUG, false) => self.writer.write_str("debug: "),
+                (Level::TRACE, false) => self.writer.write_str("trace: "),
+            };
```

Comment:
I wish there was a nicer way to do conditional coloring without writing everything out twice. This is kinda unfortunate :'-(

Reply:
78cac5fc6ab5b8facd743a10905b2063a16b4ffd should make it a bit nicer --- i forgot that there _was_ a way to do that that i had written previously, lol :/